infra/blueprint-test/README.md

```diff
@@ -0,0 +1,408 @@
+
+```go
+// The tft struct can be used to pull output variables of the TF module being invoked by this test
+op := gcloud.Run(t, fmt.Sprintf("sql instances describe %s --project %s", mySqlT.GetStringOutput("name"), mySqlT.GetStringOutput("project_id")))
+```
+
+2. Once you have retrieved values from GCP, use the (assert)[https://pkg.go.dev/github.com/stretchr/testify/assert] package to perform custom validations with respect to the resources provisioned. Here are some common assertions that can be useful in most test scenarios.
+
+   1. Contains
+
+    ```go
+    // assert values that are contained in the expected output
+    assert.Contains(op.Get("gceZone").String(), region, "GCE region is valid")
+    ```
+
+   2. Boolean (True)
+
+    ```go
+    // assert boolean values
+    assert.True(op.Get("settings.ipConfiguration.ipv4Enabled").Bool(), "ipv4 is enabled")
+    ```
+
+   3. GreaterOrEqual
+
+    ```go
+    // assert values that are greater than or equal to the expected value
+    assert.GreaterOrEqual(op.Get("settings.dataDiskSizeGb").Float(), 10.0, "at least 5 backups are retained")
+    ```
+
+   4. Empty
+
+    ```go
+    // assert values that are supposed to be empty or nil
+    assert.Empty(op.Get("settings.userLabels"), "no labels are set")
+    ```
+
+   5. Equal
+
+    ```go
+    // assert values that are supposed to be equal to the expected values
+    assert.Equal(databaseVersion, op.Get("databaseVersion").String(), "database versions is valid is set to "+databaseVersion)
+    ```
+
+The entire integration test can be found [here](https://github.com/terraform-google-modules/terraform-google-sql-db/blob/master/test/integration/mysql-public/mysql_public_test.go).
+
+# 4. Test Execution
+As mentioned in section 2 above, the blueprints test framework executes tests in two ways: auto-discovered and custom tests. Each type of test undergoes 4 stages of test execution:
+
+1. init - runs `terraform init` and `terraform validate`
+2. apply - runs `terraform apply`
+3. verify - runs `terraform plan` to verify apply successful apply with no more resources to add/destroy
+4. teardown - runs `terraform destroy`
+
+By default, tests go through 4 stages above. You can also explicitly run individual stages one at a time.
+
+## 4.1 Test prerequisites
+In order for the test to execute, certain prerequisite resources and components need to be in place. These can be set up using the TF modules under `test/setup`. Running `terraform apply` in this directory will set up all resources required for the test.
+
+```
+Note: Output values from `test/setup` are automatically loaded as Terraform environment variables and are available to both auto discovered and custom/explicit tests as inputs.
```

Comment:
It might be worth cross linking this from above where we talk about setting variable values in examples.

Reply:
Good idea. Added.